src/main/java/com/xiaomi/infra/pegasus/rpc/async/SessionFailureDetector.java

```diff
@@ -0,0 +1,40 @@
+package com.xiaomi.infra.pegasus.rpc.async;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * SessionFailureDetector detects whether the session is half-closed by the remote host, in which
+ * case we need to actively close the session and reconnect.
+ */
+public class SessionFailureDetector {
+
+  // Timestamp of the first timed out rpc.
+  private AtomicLong firstRecentTimedOutMs;
+
+  // Session is marked failure if all the RPCs across
+  // `sessionResetTimeWindowMs` are timed out.
+  public static final long FAILURE_DETECT_WINDOW_MS = 10 * 1000; // 10s
```

Comment:
maybe 10s is too short? consider set it to 30s or 60s?

Reply:
I keep it as it was in this refactoring. I will consider a larger window if it makes sense.